docs/tutorials/csharp/basic_csharp.md

```diff
@@ -1,36 +1,150 @@
+It always existed in the onnxruntime library, but was not exposed in the C# API.
+
+As before, `OrtValues` can be created directly on top of the managed `unmanaged` (struct based blittable types)  arrays.
+Read MS documentation on `blittable` data types. onnxruntime C# API allows use of managed buffers for input or output.
+
+If output shapes are known, one can pre-allocate `OrtValue` on top of the managed or unmanaged allocations and supply
+those OrtValues to be used as outputs. Due to this fact, the need for `IOBinding` is greatly diminished.
+
+String data is represented as UTF-16 string objects in C#. It will still need to be copied and converted to UTF-8 to the native
+memory. However, that conversion is now more optimized and is done in a single pass without intermediate byte arrays.
+The same applies to string `OrtValue` tensors returned as outputs. Character based API now operates on `Span<char>`,
+`ReadOnlySpan<char>`, and `ReadOnlyMemory<char>` objects. This adds flexibility to the API and allows to avoid unnecessary copies.
+
+Except some of the above deprecated API classes, nearly all of C# API classes are `IDisposable`.
+Meaning they need to be disposed after use, otherwise you will get memory leaks.
+Because OrtValues are used to hold tensor data, the sizes of the leaks can be huge. They are likely
+to accumulate with each `Run` call, as each inference call requires input OrtValues and returns output OrtValues.
+Do not hold your breath for finalizers which are not guaranteed to ever run, and if they do, they do it
+when it is too late.
+
+This includes `SessionOptions`, `RunOptions`, `InferenceSession`, `OrtValue`. Run() calls return `IDisposableCollection`
+that allows to dispose all of the containing objects in one statement or `using`. This is because these objects
+own some native resource, often a native object.
+
+Not disposing `OrtValue` that was created on top of the managed buffer would result in
+that buffer pinned in memory indefinitely. Such a buffer can not be garbage collected or moved in memory.
+
+`OrtValue`s that were created on top of the native onnxruntime memory should also be disposed of promptly.
+Otherwise, the native memory will not be deallocated. OrtValues returned by `Run()` usually hold native memory.
+
+GC can not operate on native memory or any other native resources.
+
+The `using` statement or a block is a convenient way to ensure that the objects are disposed.
+`InferenceSession` can be a long lived object and a member of another class. It eventually must also need to be disposed.
+This means, the containing class also would have to be made disposable to achieve this.
+
+OrtValue API also provides visitor like API to walk ONNX maps and sequences.
+This is a more efficient way to access Onnxruntime data.
 
 To start scoring using the model, open a session using the `InferenceSession` class, passing in the file path to the model as a parameter.
```

Comment:
`Run a model with C#` heading?

Reply:
Instead of `open a session`, say `instantiate InferenceSession class`